<a href="https://colab.research.google.com/github/cysorianoc/Hugging_Face/blob/main/Notebook_6_Text_to_Speech.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Lesson 7: Text to Speech

Text to speech is a one to many problem, in classification there are many (inifinite) ways to say a sentence. Each person has a different way of speaking and all forms are correct.

There are open source modelas as ```vlts``` from kakao enterprise and that can handle this task.

- In the classroom, the libraries are already installed for you.
- If you would like to run this code on your own machine, you can install the following:

```
    !pip install transformers
    !pip install gradio
    !pip install timm
    !pip install timm
    !pip install inflect
    !pip install phonemizer
    
```

**Note:**  `py-espeak-ng` is only available Linux operating systems.

To run locally in a Linux machine, follow these commands:
```
    sudo apt-get update
    sudo apt-get install espeak-ng
    pip install py-espeak-ng
```

In [1]:
    !pip install transformers
    !pip install gradio
    !pip install timm
    !pip install timm
    !pip install inflect
    !pip install phonemizer

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 48.4/48.4 kB 2.1 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 108.5/108.5 kB 5.0 MB/s eta 0:00:00


In [3]:
!apt-get update
!apt-get install espeak-ng
!pip install py-espeak-ng

Get:1 https://cli.github.com/packages stable InRelease [3,917 B]
Get:2 http://security.ubuntu.com/ubuntu jammy-security InRelease [129 kB]
Get:3 https://cloud.r-project.org/bin/linux/ubuntu jammy-cran40/ InRelease [3,632 B]
Get:4 https://r2u.stat.illinois.edu/ubuntu jammy InRelease [6,555 B]
Hit:5 http://archive.ubuntu.com/ubuntu jammy InRelease
Get:6 http://archive.ubuntu.com/ubuntu jammy-updates InRelease [128 kB]
Get:7 https://cli.github.com/packages stable/main amd64 Packages [356 B]
Get:8 https://ppa.launchpadcontent.net/deadsnakes/ppa/ubuntu jammy InRelease [18.1 kB]
Get:9 https://cloud.r-project.org/bin/linux/ubuntu jammy-cran40/ Packages [105 kB]
Hit:10 https://ppa.launchpadcontent.net/ubuntugis/ppa/ubuntu jammy InRelease
Get:11 http://archive.ubuntu.com/ubuntu jammy-backports InRelease [127 kB]
Get:12 https://r2u.stat.illinois.edu/ubuntu jammy/main all Packages [10.7 MB]
Get:13 http://security.ubuntu.com/ubuntu jammy-security/main amd64 Packages [4,190 kB]
Get:14 https://r2u.s

### Build the `text-to-speech` pipeline using the 🤗 Transformers Library

- Here is some code that suppresses warning messages.

In [4]:
from transformers.utils import logging

logging.set_verbosity_error()

In [5]:
from transformers import pipeline

narrator = pipeline("text-to-speech",
                    model="kakao-enterprise/vits-ljs")

config.json:   0%|          | 0.00/1.64k [00:00<?, ?B/s]

model.safetensors: reconstructing file:   0%|          |  0.00B /  145MB            

model.safetensors: downloading bytes:           |  0.00B            

Loading weights:   0%|          | 0/762 [00:00<?, ?it/s]

tokenizer_config.json:   0%|          | 0.00/285 [00:00<?, ?B/s]

vocab.json:   0%|          | 0.00/2.14k [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/47.0 [00:00<?, ?B/s]

Info about [kakao-enterprise/vits-ljs](https://huggingface.co/kakao-enterprise/vits-ljs)

In [6]:
text = """
Researchers at the Allen Institute for AI, \
HuggingFace, Microsoft, the University of Washington, \
Carnegie Mellon University, and the Hebrew University of \
Jerusalem developed a tool that measures atmospheric \
carbon emitted by cloud servers while training machine \
learning models. After a model’s size, the biggest variables \
were the server’s location and time of day it was active.
"""

In [7]:
narrated_text = narrator(text)

In [9]:
from IPython.display import Audio as IPythonAudio

IPythonAudio(narrated_text["audio"],
             rate=narrated_text["sampling_rate"])

There are two variants of the VITS model: one is trained on the LJ Speech dataset, and the other is trained on the VCTK dataset. LJ Speech dataset consists of 13,100 short audio clips of a single speaker with a total length of approximately 24 hours. The VCTK dataset consists of approximately 44,000 short audio clips uttered by 109 native English speakers with various accents. The total length of the audio clips is approximately 44 hours.

Let's see the result with ```vits-vctk``` variant

In [10]:
from transformers import pipeline

pipe = pipeline("text-to-speech", model="kakao-enterprise/vits-vctk")

config.json:   0%|          | 0.00/1.65k [00:00<?, ?B/s]

model.safetensors: reconstructing file:   0%|          |  0.00B /  159MB            

model.safetensors: downloading bytes:           |  0.00B            

Loading weights:   0%|          | 0/782 [00:00<?, ?it/s]

tokenizer_config.json:   0%|          | 0.00/285 [00:00<?, ?B/s]

vocab.json:   0%|          | 0.00/2.14k [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/47.0 [00:00<?, ?B/s]

In [11]:
narrated_text=pipe(text)

In [12]:
from IPython.display import Audio as IPythonAudio

IPythonAudio(narrated_text["audio"],
             rate=narrated_text["sampling_rate"])

Finally let's try ```wfloat``` model.

wfloat-tts is a lightweight multi-speaker English VITS text-to-speech model with speaker, emotion, and intensity control


[tSource](https://huggingface.co/Wfloat/wfloat-tts)

In [24]:
!pip install wfloat

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 45.5/45.5 kB 2.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 10.1/10.1 MB 17.4 MB/s eta 0:00:00


In [27]:
import wfloat
from IPython.display import Audio as IPythonAudio

model = wfloat.load("wfloat/wfloat-tts")

result = model.generate(
    text="Hey there, how are you today?",
    voice_id="narrator_woman",
    emotion="neutral",
    intensity=0.5,
)

# Save to WAV
result.audio.save("out.wav")

# Play audio in the notebook
IPythonAudio(
    result.audio.samples,
    rate=result.audio.sample_rate
)

In [30]:
text_test = """
Researchers at the Allen Institute for AI, \
HuggingFace, Microsoft, the University of Washington, \
Carnegie Mellon University, and the Hebrew University of \
Jerusalem developed a tool that measures atmospheric \
carbon emitted by cloud servers while training machine \
learning models. After a model’s size, the biggest variables \
were the server’s location and time of day it was active.
"""

In [33]:
import wfloat
from IPython.display import Audio as IPythonAudio

model = wfloat.load("wfloat/wfloat-tts")

result = model.generate(
    text=text_test,
    voice_id="narrator_man",
    emotion="neutral",
    intensity=0.5,
)

# Save to WAV
result.audio.save("out.wav")

# Play audio in the notebook
IPythonAudio(
    result.audio.samples,
    rate=result.audio.sample_rate
)